### DISCUSSION TITLE
Drop Clearcase support? [ANSWERED - NO]

### DISCUSSION THREAD

**Discussion by eteran** [eteran/nedit-ng#230] [General]
@sjtringali @tksoh (and anyone with an opinion!)

I'm a big believer in less code = fewer bugs. So when I see a feature that seemingly isn't used anymore, my instinct is to think about just deleting it.

So the question is in the title :-). Should we just cut Clearcase support? 

I've never used it, and frankly, if I were to make changes any more complex than what I've already done, I'd have no way of testing that code to ensure that it is correct. Is it worth keeping around at all?

**Comment by anjohnson** [comment]
No objections here, what did it do? Nowadays some kind of integration with Git would make more sense, but I wouldn't personally put that very high on any kind of a priority list. That would seem to fall more into plugin territory IMHO.

**Comment by tksoh** [comment]
I'd like to repeat what I said on NG replacing nedit 5: we don't know which features are [not] being used. According to wikipedia, ClearCase is not dead:

> V9.1.0 (December 2020):
> See what's new in IBM Rational ClearCase 9.1.0.

If we can turn it into a plugin (or shell commands?), that would be a good compromise. And while we are there, we should also add support for other VCS tool also: CVS (huh!?), git, Mercurial, SVN, etc

**Reply by eteran** [reply]
Good points!

**Reply by tksoh** [reply]
@eteran I'd open a feature request on the issue tracker on adding the other VCS, just as a reminder :-)

**Reply by eteran** [reply]
Sure, that makes sense. We also should clarify what "git support" really means for us.

Is it a context menu that will let you commit? Does it let you pull? Push? Merge? Etc...

**Reply by tksoh** [reply]
For starter, I think we should add the basic commands that can be achieved zero/minimum UI work. Perhaps we should look at how other people do it first before we dive further. But one thing is clear, most modern VCS already have their their own Tortoise* GUI, we probably don't need to try replacing them, so we should limit ourselves to a frequently used subset of their functions.

On the menu, VCS is so crucial these days that it might warrant it's own menu. If we want to preserve the current appearance of nedit, then we can perhaps add it under the Edit menu.

I also like how PyCharm handles it by adding in on the status bar (yes, we will need a status bar). I think it's quite intuitive as it also show which VCS is being used. Though we would probably still need the VCS menu to advertise it's presence.

**Comment by sjtringali** [comment]
Agree. Any VCS integration is best done with macros or plugins.

Clearance is special because it is a virtual file system and can’t be easily done with macros. 

Honestly unless we have a way to test it, it could break easily. But, I say leave it in - if someone notices it’s broken then we have tester. As long as it’s not taking up large amounts of UI should be fine.

**Comment by sjtringali** [comment]
There is something to be said for small inbuilt UI things, mostly status and name things that are read-only like the window title. So I can buy in to native git support that shows status information *directly* related to the file being edited. Like: if the file differs from the index or HEAD, etc.

Something like “git log”, for example, isn’t so good, because now it’s shoehorning an entirely different concept (metadata about the repository as a whole) into a file editor. It’s an impedance mismatch in my mind, best handled by other tools.

As for mutating actions, it’s an unbounded set of things... as above. these are better handled by (preferably) macros or (less so) plugins. Macros are customizable by the user, and despite being old, the utility is proven. If anything, adding more macro primitives is the way to go. That’s a plug-in system I could get behind.

I was up and running with nedit in perforce in a minute or two by editing trivial shell macros; people trying to customize IDEs with compiled plugins? Quite a bit longer.

**Comment by eteran** [comment]
Ah, a git plugin that adds to the status bar something like the current branch and an icon reflecting the "dirty" status seems VERY useful. I like this idea a lot!

**Comment by tksoh** [comment]
> Macros are customizable by the user, and despite being old, the utility is proven. If anything, adding more macro primitives is the way to go. That’s a plug-in system I could get behind.

I am not sure what you have in mind, but I like the way it sounds. 

If we can add the support for VCS to the macros, which allows some level of interaction with nedit in order to display common info like "dirty" status, branch name, HEAD commit info, etc., then we can leave to users to add the ones they are using. And we can bundle the common ones ourselves.

For things like 'git log', we probably just call 'gitk', if it's available.
